core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -322,7 +325,9 @@ void register(MutableRel result, MutableRel query) {
               ImmutableList.of(condition2, target2));
       RexNode r =
           canonizeNode(rexBuilder, simplify.simplifyUnknownAsFalse(x2));
-      if (!r.isAlwaysFalse() && isEquivalent(condition2, r)) {
+      RexNode simplifiedCond2 =
+          canonizeNode(rexBuilder, simplify.simplifyUnknownAsFalse(condition2));
+      if (!r.isAlwaysFalse() && isEquivalent(simplifiedCond2, r)) {
```

Comment:
I agree with Early Return, it cuts out some unnecessary steps. But doing another simplify for condition2 seems unnecessary, condition2 only canonizeNode after simplify.
How about using simplify instead of simplifyUnknownAsFalse for _x2_ .

Reply:
Actually I'm not sure why this function uses both simplify and simplifyUnknownAsFalse, so I made a conservative change to minimize impact. If possible, we could replace simplify with simplifyUnknownAsFalse, but that seems too radical.

Reply:
I agree with @suibianwanwank, we shouldn't need another simplification on `condition2` while early return seems good.

Here we are dealing with filtering predicates, so I guess that `simplifyUnknownAsFalse` has the right semantics, rather than just `simplify`, but we need tests covering these changes and showing that we are doing the right thing on corner cases involving `unknown` and the early return condition.

@hannerwang, can you add them?

Reply:
Hello @asolimando ,
I attempted to change simplifyUnknownAsFalse to simplify, but I encountered some issues. 
For instance, with the expression `x >= 5 and x > 5`, the simplify method does not reduce it, whereas simplifyUnknownAsFalse simplifies it to` x > 5`. Additionally, expressions like` x > 5 and x < 3` are simplified to `x <> x` when UNKNOWN is treated as UNKNOWN, and to `false` when UNKNOWN is treated as FALSE. 
Therefore, we cannot simply replace simplifyUnknownAsFalse with simplify, as it would cause many tests to fail.